examples/petsc/src/petscutils.c

```diff
@@ -125,28 +125,56 @@ PetscErrorCode SetupDMByDegree(DM dm, PetscInt degree, PetscInt num_comp_u,
   PetscInt ierr, marker_ids[1] = {1};
   PetscFE fe;
   MPI_Comm comm;
+  PetscBool      is_simplex = PETSC_TRUE;
 
   PetscFunctionBeginUser;
 
+  // Check if simplex or tensor-product mesh
+  ierr = DMPlexIsSimplex(dm, &is_simplex); CHKERRQ(ierr);
   // Setup FE
   ierr = PetscObjectGetComm((PetscObject)dm, &comm); CHKERRQ(ierr);
-  ierr = PetscFECreateLagrange(comm, dim, num_comp_u, PETSC_FALSE, degree, degree,
+  ierr = PetscFECreateLagrange(comm, dim, num_comp_u, is_simplex, degree, degree,
                                &fe); CHKERRQ(ierr);
   ierr = DMSetFromOptions(dm); CHKERRQ(ierr);
   ierr = DMAddField(dm, NULL, (PetscObject)fe); CHKERRQ(ierr);
-  {
-    /* create FE field for coordinates */
+  ierr = DMCreateDS(dm); CHKERRQ(ierr);
+  if (0) {
+    DM             dm_coord;
+    PetscDS        ds_coord;
+    PetscFE        fe_coord_current, fe_coord_new;
+    PetscDualSpace fe_coord_dual_space;
+    PetscInt       fe_coord_order, num_comp_coord;
+
+    ierr = DMGetCoordinateDM(dm, &dm_coord); CHKERRQ(ierr);
+    ierr = DMGetCoordinateDim(dm, &num_comp_coord); CHKERRQ(ierr);
+    ierr = DMGetRegionDS(dm_coord, NULL, NULL, &ds_coord); CHKERRQ(ierr);
+    ierr = PetscDSGetDiscretization(ds_coord, 0, (PetscObject *)&fe_coord_current);
+    CHKERRQ(ierr);
+    ierr = PetscFEGetDualSpace(fe_coord_current, &fe_coord_dual_space);
+    CHKERRQ(ierr);
+    ierr = PetscDualSpaceGetOrder(fe_coord_dual_space, &fe_coord_order);
+    CHKERRQ(ierr);
+
+    // Create FE for coordinates
+    ierr = PetscFECreateLagrange(comm, dim, num_comp_coord, is_simplex,
+                                 fe_coord_order, degree, &fe_coord_new);
+    CHKERRQ(ierr);
+    ierr = DMProjectCoordinates(dm, fe_coord_new); CHKERRQ(ierr);
+    ierr = PetscFEDestroy(&fe_coord_new); CHKERRQ(ierr);
+  }
+  if (1) {
+    // create FE field for coordinates
     PetscFE fe_coords;
     PetscInt num_comp_coord;
     ierr = DMGetCoordinateDim(dm, &num_comp_coord); CHKERRQ(ierr);
-    ierr = PetscFECreateLagrange(comm, dim, num_comp_coord, PETSC_FALSE, 1, 1,
+    ierr = PetscFECreateLagrange(comm, dim, num_comp_coord, is_simplex, 1, degree,
                                  &fe_coords); CHKERRQ(ierr);
     ierr = DMProjectCoordinates(dm, fe_coords); CHKERRQ(ierr);
     ierr = PetscFEDestroy(&fe_coords); CHKERRQ(ierr);
   }
 
```

Comment:
@jedbrown @knepley @jeremylt I get the error in the first block. In both blocks we try to set the `fe_coord` of the `dm` created by `DMPlexCreateBoxMesh`. I wanted to check if `PetscDualSpaceGetOrder` returns `1` for tensor basis or not, but I get error. The order in second block is set to `1` manually.  I get the following error when I turn on the first block
```
valgrind --leak-check=full ./bps -problem bp1 -degree 3 -dm_view
==501897== Memcheck, a memory error detector
==501897== Copyright (C) 2002-2017, and GNU GPL'd, by Julian Seward et al.
==501897== Using Valgrind-3.15.0 and LibVEX; rerun with -h for copyright info
==501897== Command: ./bps -problem bp1 -degree 3 -dm_view
==501897== 
==501897== Invalid read of size 8
==501897==    at 0x4FCE854: PetscFEGetDualSpace (in /home/rezgar/petsc/arch-mpich/lib/libpetsc.so.3.016.4)
==501897==    by 0x114A91: SetupDMByDegree (in /home/rezgar/libCEED/examples/petsc/bps)
==501897==    by 0x10DD0D: main (in /home/rezgar/libCEED/examples/petsc/bps)
==501897==  Address 0xd9a41f8 is 8 bytes after a block of size 32 alloc'd
==501897==    at 0x483E340: memalign (in /usr/lib/x86_64-linux-gnu/valgrind/vgpreload_memcheck-amd64-linux.so)
==501897==    by 0x4959EEA: PetscMallocAlign (in /home/rezgar/petsc/arch-mpich/lib/libpetsc.so.3.016.4)
==501897==    by 0x495A897: PetscMallocA (in /home/rezgar/petsc/arch-mpich/lib/libpetsc.so.3.016.4)
==501897==    by 0x52F9EE9: DMFieldEnlarge_Static.isra.0 (in /home/rezgar/petsc/arch-mpich/lib/libpetsc.so.3.016.4)
==501897==    by 0x5305341: DMAddField (in /home/rezgar/petsc/arch-mpich/lib/libpetsc.so.3.016.4)
==501897==    by 0x530555E: DMSetNumFields (in /home/rezgar/petsc/arch-mpich/lib/libpetsc.so.3.016.4)
==501897==    by 0x5176D4C: DMCreateCoordinateDM_Plex (in /home/rezgar/petsc/arch-mpich/lib/libpetsc.so.3.016.4)
==501897==    by 0x5313DD1: DMGetCoordinateDM (in /home/rezgar/petsc/arch-mpich/lib/libpetsc.so.3.016.4)
==501897==    by 0x5313F35: DMGetCoordinateSection (in /home/rezgar/petsc/arch-mpich/lib/libpetsc.so.3.016.4)
==501897==    by 0x515CB0D: DMPlexCreateBoxMesh_Internal (in /home/rezgar/petsc/arch-mpich/lib/libpetsc.so.3.016.4)
==501897==    by 0x515DC3D: DMPlexCreateBoxMesh (in /home/rezgar/petsc/arch-mpich/lib/libpetsc.so.3.016.4)
==501897==    by 0x10DB55: main (in /home/rezgar/libCEED/examples/petsc/bps)
==501897== 
==501897== Invalid read of size 4
==501897==    at 0x4FF9594: PetscDualSpaceGetOrder (in /home/rezgar/petsc/arch-mpich/lib/libpetsc.so.3.016.4)
==501897==    by 0x114AA8: SetupDMByDegree (in /home/rezgar/libCEED/examples/petsc/bps)
==501897==    by 0x10DD0D: main (in /home/rezgar/libCEED/examples/petsc/bps)
==501897==  Address 0x2a8 is not stack'd, malloc'd or (recently) free'd
==501897== 
[0]PETSC ERROR: ------------------------------------------------------------------------
[0]PETSC ERROR: Caught signal number 11 SEGV: Segmentation Violation, probably memory access out of range
[0]PETSC ERROR: Try option -start_in_debugger or -on_error_attach_debugger
[0]PETSC ERROR: or see https://petsc.org/release/faq/#valgrind
[0]PETSC ERROR: or try http://valgrind.org on GNU/linux and Apple Mac OS X to find memory corruption errors
[0]PETSC ERROR: configure using --with-debugging=yes, recompile, link, and run 
[0]PETSC ERROR: to get more information on the crash.
[0]PETSC ERROR: --------------------- Error Message --------------------------------------------------------------
[0]PETSC ERROR: Signal received
[0]PETSC ERROR: See https://petsc.org/release/faq/ for trouble shooting.
[0]PETSC ERROR: Petsc Development GIT revision: v3.16.4-838-g2d5b27b880  GIT Date: 2022-02-10 17:33:07 +0000
[0]PETSC ERROR: ./bps on a arch-mpich named rezgar by rezgar Wed Feb 16 13:45:44 2022
[0]PETSC ERROR: Configure options --download-ctetgen --download-mumps --download-scalapack --download-triangle --with-debugging=0 COPTFLAGS="-O3 -march=native" PETSC_ARCH=arch-mpich
[0]PETSC ERROR: #1 User provided function() at unknown file:0
[0]PETSC ERROR: Run with -malloc_debug to check if memory corruption is causing the crash.
--------------------------------------------------------------------------
MPI_ABORT was invoked on rank 0 in communicator MPI_COMM_WORLD
with errorcode 59.

NOTE: invoking MPI_ABORT causes Open MPI to kill all MPI processes.
You may or may not see output from other processes, depending on
exactly when Open MPI kills them.
```

Reply:
First compile debugging so we can see line numbers. It seems like you are not passing an FE.

Reply:
Hmm, I get this
```
[0]PETSC ERROR: --------------------- Error Message --------------------------------------------------------------
[0]PETSC ERROR: Invalid argument
[0]PETSC ERROR: Wrong type of object: Parameter # 1
[0]PETSC ERROR: See https://petsc.org/release/faq/ for trouble shooting.
[0]PETSC ERROR: Petsc Development GIT revision: v3.16.4-838-g2d5b27b880  GIT Date: 2022-02-10 17:33:07 +0000
[0]PETSC ERROR: ./bps on a arch-mpich named rezgar by rezgar Wed Feb 16 14:44:49 2022
[0]PETSC ERROR: Configure options --download-ctetgen --download-mumps --download-scalapack --download-triangle --with-debugging=1 COPTFLAGS="-O3 -march=native" PETSC_ARCH=arch-mpich
[0]PETSC ERROR: #1 PetscFEGetDualSpace() at /home/rezgar/petsc/src/dm/dt/fe/interface/fe.c:592
[0]PETSC ERROR: #2 SetupDMByDegree() at /home/rezgar/libCEED/examples/petsc/src/petscutils.c:154
[0]PETSC ERROR: #3 Run() at /home/rezgar/libCEED/examples/petsc/bps.c:441
[0]PETSC ERROR: #4 main() at /home/rezgar/libCEED/examples/petsc/bps.c:601
[0]PETSC ERROR: PETSc Option Table entries:
[0]PETSC ERROR: -degree 3
[0]PETSC ERROR: -dm_view
[0]PETSC ERROR: -problem bp1
```

Reply:
You are not passing a PetscFE

Reply:
Thanks, I'll check what I missed.

Reply:
Makes me think we got the wong ds or ds field on the lines above

Reply:
Yeah, I need to check ratel to see what I missed since I copied the first block from there.